File: Examples/haskell/sorting.hs
-- Sorting algorithms in Haskell
-- Quicksort, merge sort, insertion sort

module Main where

import Data.List (intercalate)

-- ── Quicksort ──────────────────────────────────────────────────────────
quicksort :: Ord a => [a] -> [a]
quicksort [] = []
quicksort (x:xs) =
  quicksort smaller ++ [x] ++ quicksort larger
  where
    smaller = filter (<= x) xs
    larger  = filter (>  x) xs

-- ── Merge sort ────────────────────────────────────────────────────────
mergesort :: Ord a => [a] -> [a]
mergesort [] = []
mergesort [x] = [x]
mergesort xs  = merge (mergesort left) (mergesort right)
  where
    (left, right) = splitAt (length xs `div` 2) xs
    merge [] ys = ys
    merge xs [] = xs
    merge (x:xs) (y:ys)
      | x <= y    = x : merge xs (y:ys)
      | otherwise = y : merge (x:xs) ys

-- ── Insertion sort ────────────────────────────────────────────────────
insertionsort :: Ord a => [a] -> [a]
insertionsort = foldr insert []
  where
    insert x [] = [x]
    insert x (y:ys)
      | x <= y    = x : y : ys
      | otherwise = y : insert x ys

-- ── Show results ──────────────────────────────────────────────────────
showList' :: Show a => [a] -> String
showList' xs = "[" ++ intercalate ", " (map show xs) ++ "]"

main :: IO ()
main = do
  let unsorted = [64, 25, 12, 22, 11, 90, 3, 77] :: [Int]
  putStrLn "=== Haskell Sorting Algorithms ==="
  putStrLn $ "Unsorted:       " ++ showList' unsorted
  putStrLn $ "Quicksort:      " ++ showList' (quicksort unsorted)
  putStrLn $ "Merge sort:     " ++ showList' (mergesort unsorted)
  putStrLn $ "Insertion sort: " ++ showList' (insertionsort unsorted)

  putStrLn "\n=== Sorting strings ==="
  let words' = ["banana", "apple", "cherry", "date", "elderberry"]
  putStrLn $ "Unsorted: " ++ showList' words'
  putStrLn $ "Sorted:   " ++ showList' (quicksort words')

File: Examples/prolog/sorting.pl
% Prolog Sorting Algorithms
% Quicksort, merge sort, insertion sort, and bubble sort

% ── Quicksort ──────────────────────────────────────────────────────────
partition(_, [], [], []).
partition(Pivot, [H|T], [H|Less], Greater) :-
    H =< Pivot, !,
    partition(Pivot, T, Less, Greater).
partition(Pivot, [H|T], Less, [H|Greater]) :-
    partition(Pivot, T, Less, Greater).

quicksort([], []).
quicksort([H|T], Sorted) :-
    partition(H, T, Less, Greater),
    quicksort(Less, SortedLess),
    quicksort(Greater, SortedGreater),
    append(SortedLess, [H|SortedGreater], Sorted).

% ── Merge sort ─────────────────────────────────────────────────────────
split([], [], []).
split([X], [X], []).
split([X,Y|T], [X|L], [Y|R]) :- split(T, L, R).

merge([], R, R).
merge(L, [], L).
merge([H1|T1], [H2|T2], [H1|Merged]) :-
    H1 =< H2, !,
    merge(T1, [H2|T2], Merged).
merge([H1|T1], [H2|T2], [H2|Merged]) :-
    merge([H1|T1], T2, Merged).

mergesort([], []).
mergesort([X], [X]).
mergesort(List, Sorted) :-
    List = [_,_|_],
    split(List, L, R),
    mergesort(L, SL),
    mergesort(R, SR),
    merge(SL, SR, Sorted).

% ── Insertion sort ─────────────────────────────────────────────────────
insert_sorted(X, [], [X]).
insert_sorted(X, [H|T], [X,H|T]) :- X =< H, !.
insert_sorted(X, [H|T], [H|R]) :- insert_sorted(X, T, R).

insertion_sort([], []).
insertion_sort([H|T], Sorted) :-
    insertion_sort(T, ST),
    insert_sorted(H, ST, Sorted).

% ── Bubble sort ────────────────────────────────────────────────────────
bubble_pass([], [], false).
bubble_pass([X], [X], false).
bubble_pass([X,Y|T], [X|R], Swapped) :-
    X =< Y, !,
    bubble_pass([Y|T], R, Swapped).
bubble_pass([X,Y|T], [Y|R], true) :-
    bubble_pass([X|T], R, _).

bubble_sort(L, Sorted) :- bubble_sort_(L, Sorted, false).
bubble_sort_(L, L, false) :- !.  % no swaps → already sorted
bubble_sort_(L, Sorted, _) :-
    bubble_pass(L, L1, _),
    bubble_sort(L1, Sorted).

% ── Demos ──────────────────────────────────────────────────────────────
:- initialization(main).

main :-
    writeln('=== Prolog Sorting Algorithms ==='),
    Data = [64, 25, 12, 90, 3, 77, 44, 18, 55, 37],
    write('Input: '), writeln(Data),
    nl,

    quicksort(Data, QS),
    write('Quicksort:      '), writeln(QS),

    mergesort(Data, MS),
    write('Merge sort:     '), writeln(MS),

    insertion_sort(Data, IS),
    write('Insertion sort: '), writeln(IS),

    bubble_sort(Data, BS),
    write('Bubble sort:    '), writeln(BS),

    nl,
    writeln('=== Sorted string list ==='),
    Words = [banana, apple, cherry, date, elderberry],
    write('Input: '), writeln(Words),
    quicksort(Words, SortedWords),
    write('Sorted: '), writeln(SortedWords).
